src/block/inventory/EnchantInventory.php

```diff
@@ -23,18 +23,83 @@
 
 namespace pocketmine\block\inventory;
 
+use pocketmine\event\inventory\PlayerEnchantmentOptionsRequestEvent;
 use pocketmine\inventory\SimpleInventory;
 use pocketmine\inventory\TemporaryInventory;
+use pocketmine\item\enchantment\EnchantmentHelper as Helper;
+use pocketmine\item\enchantment\EnchantmentOption;
+use pocketmine\item\Item;
+use pocketmine\item\ItemTypeIds;
+use pocketmine\item\VanillaItems;
 use pocketmine\world\Position;
+use function array_keys;
+use function array_search;
+use function count;
 
 class EnchantInventory extends SimpleInventory implements BlockInventory, TemporaryInventory{
 	use BlockInventoryTrait;
 
 	public const SLOT_INPUT = 0;
 	public const SLOT_LAPIS = 1;
 
+	/** @var EnchantmentOption[] $options */
+	private array $options = [];
+
 	public function __construct(Position $holder){
 		$this->holder = $holder;
 		parent::__construct(2);
 	}
+
+	protected function onSlotChange(int $index, Item $before) : void{
+		if($index === self::SLOT_INPUT){
+			foreach($this->viewers as $viewer){
+				$this->options = [];
+				$item = $this->getInput();
+				$options = Helper::getEnchantOptions($this->holder, $item, $viewer->getXpSeed());
+
+				$event = new PlayerEnchantmentOptionsRequestEvent($viewer, $this, $options);
+				$event->call();
+				if(!$event->isCancelled() && count($event->getOptions()) > 0){
+					foreach($event->getOptions() as $option){
+						$this->options[$option->getNetworkId()] = $option;
+					}
+					$viewer->getNetworkSession()->sendEnchantOptions($this->options);
+				}
+			}
+		}
+
+		parent::onSlotChange($index, $before);
+	}
+
+	public function getInput() : Item{
+		return $this->getItem(self::SLOT_INPUT);
+	}
+
+	public function getLapis() : Item{
+		return $this->getItem(self::SLOT_LAPIS);
+	}
+
+	public function getOutput(int $optionId) : ?Item{
+		$option = $this->options[$optionId] ?? null;
+		if($option === null){
+			// Failed to find an enchantment option with the passed network id
+			return null;
+		}
+
+		$outputItem = $this->getInput();
+		if($outputItem->getTypeId() === ItemTypeIds::BOOK){
+			$outputItem = VanillaItems::ENCHANTED_BOOK();
+		}
+
+		foreach($option->getEnchantments() as $enchantment){
+			$outputItem->addEnchantment($enchantment);
+		}
+
+		return $outputItem;
+	}
+
+	public function getOptionEnchantmentLevel(int $optionId) : ?int{
+		$level = array_search($optionId, array_keys($this->options), true);
+		return $level === false ? null : $level + 1;
```

Comment:
optionId is not equal to the enchantment level. More precisely, in our enchantment system, of course, it is equal (more precisely, it is optionId + 1), because I send 0, 1 and 2 as optionId. But plugins can set other values of optionId (`networkId` in EnchantmentOption). And the enchantment level is determined by the order in which the options are added to the array.